vslib/src/sai_vs_fdb.cpp

```diff
@@ -328,18 +338,21 @@ sai_status_t internal_vs_flush_fdb_entries(
         attrs[0].value.s32 = SAI_FDB_ENTRY_TYPE_DYNAMIC;
 
         meta_sai_on_fdb_event(1, &data); // update metadata
-
+        
+        /* notification has been sent for individual mac delete, so 
+         * not sending the bulk flush notification
+         */
+#if 0    
```

Comment:
this is still an issue, notification callback must be called if it was set

Reply:
As per my understanding, notification sent here if for syncd and meta_sai_on_fdb_event() is to update local SAI meta data (meta_sai_on_fdb_event(1, &data); // update metadata). The notification is already being sent in sendFdbDelNotification(), sending it again will cause duplicate notifications to syncd. Is my understanding correct?

Reply:
Actually, the new function sendFdbDelNotification() that I added, calls meta_sai_on_fdb_event() for individual mac and then send notification. So, we don't have to do it again. My previous change was correct. Please take a look at this in at end of function internal_vs_flush_fdb_entries():

fi.bridge_port_id = it->second.at("SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID")->getAttr()->value.oid;
SWSS_LOG_INFO("FDBDelNotification: fdb_entry:%s bpid:%s", sai_serialize_fdb_entry(fi.fdb_entry).c_str(), sai_serialize_object_id(fi.bridge_port_id).c_str());
it = fdbs.erase(it);
sendFdbDelNotification(fi);

And in function sendFdbDelNotification():

    // update metadata DB
    meta_sai_on_fdb_event(1, &data)

    if (ntf != NULL)
    {
        ntf(1, &data);
    }

I will change the code back to my previous change.

Reply:
still heave some doubts on that, major change is that in previous implementation flush notifications are consolidated into "1" notifications, for example if you flash for all for 1 specific bridge id (static/dynamic) you willl get only 1 notification (even for 1000 macs), which is correct (consolidation is correct, take a look at fdb headers in SAI repo about flush api comments) and here your change is adding 1 notification per 1 single entry, which will flood receiver and we probably want to avoid that

Reply:
> still heave some doubts on that, major change is that in previous implementation flush notifications are consolidated into "1" notifications, for example if you flash for all for 1 specific bridge id (static/dynamic) you willl get only 1 notification (even for 1000 macs), which is correct (consolidation is correct, take a look at fdb headers in SAI repo about flush api comments) and here your change is adding 1 notification per 1 single entry, which will flood receiver and we probably want to avoid that

Hi Kamil, if though VS sent one notification for all mac, syncd does not support the cases where mac are flushed on either one bridge port or one vlan (per port, per vlan or per port-vlan). 
In function redisPutFdbEntryToAsicView(), 

        else if (port_oid && !bv_id)
        {
            /* we got a flush port fdb event here        */
            /* not supported yet, this is a place holder */
            /* example of a flush port fdb event         */
            /*
        }
        else if (!port_oid && bv_id)
        {
            /* we got a flush vlan fdb event here        */
            /* not supported yet, this is a place holder */
            /* example of a flush vlan event             */
            /*
         }
        else
        {
            SWSS_LOG_ERROR("received a flush fdb event, port_oid = 0x%lx, bv_id = 0x%lx, unsupported", port_oid, bv_id);
        }

Even if we try to handle this consolidated message in syncd, it will require reading the data of all the mac entries and match the bridge port or vlan and then delete it, which is a very very slow operation. 
Sending 1 notification per mac from VS can be handled in syncd irrespective of all/per port, per vlan and per port-vlan flush. It is same as what is being done from broadcom SAI, for an example.

Reply:
its not matter of supporting this on syncd, syncd that should be supported on vendor SAI to delete all mac's. Syncd should only support removing those from redis DB, and as this im not aware if it supported since this part code you pointers is not code added by me. But i believe both scenarios should be supported anyway, consolidated and 1 by 1. If you really want to send this from VS 1 by 1, please add option on virtual switch profile so that can be switched on/off allowing user to choose which scenario should be uses. In our case we are not dealing as much L2 so its not issues for us.

Reply:
also, if you are pointing out that this is expensive operation, then  matching 1 by 1 also can be slower operation, for example you could receive 1000 notifications to remove 1k fdb entries, and each of them would need to be processed separately, in consolidated message, probably a lua scrip would be better to achieve goal since just getting all mac addresses and them removing them could indeed be slow

Reply:
Hi Kamil, yes LUA script seems to be the best option to follow to handle consolidated event, I will work on that. For VS, it seems to be better than 1 by 1 event. 
For broadcom SAI 1 by 1 seems to be better as even though processing takes time, syncd notification thread is not blocked.

Reply:
thanks, brcm sai internally sends event 1 by 1, so there is no workaround here

Reply:
I reverted to single notification and added support for all types of flush in syncd using lua script.
The lua script is in the following pull request:
https://github.com/Azure/sonic-swss-common/pull/303